dm/worker/server.go

```diff
@@ -136,115 +134,114 @@ func (s *Server) Close() {
-		err = s.worker.PauseSubTask(name)
+		opLogID, err = s.worker.PauseSubTask(name)
 	case pb.TaskOp_Resume:
-		err = s.worker.ResumeSubTask(name)
+		opLogID, err = s.worker.ResumeSubTask(name)
 	default:
-		resp.Msg = fmt.Sprintf("invalid operate %s on sub task", req.Op.String())
-		return resp, nil
+		return nil, errors.Errorf("invalid operate %s on sub task", req.Op.String())
 	}
-
 	if err != nil {
-		log.Errorf("[server] operate(%s) sub task %s error %v", req.Op.String(), name, errors.ErrorStack(err))
-		resp.Msg = errors.ErrorStack(err)
-		return resp, nil
-	}
-
-	resp.Result = true
-	return resp, nil
+		err = errors.Annotatef(err, "operate(%s) sub task %s", req.Op.String(), name)
+		log.Errorf("[server] %s", errors.ErrorStack(err))
+		return nil, err
+	}
+
+	return &pb.OperateSubTaskResponse{
+		Meta: &pb.CommonWorkerResponse{
+			Result: true,
+			Msg:    "",
+		},
+		Op:    req.Op,
+		LogID: opLogID,
+	}, nil
 }
 
 // UpdateSubTask implements WorkerServer.UpdateSubTask
-func (s *Server) UpdateSubTask(ctx context.Context, req *pb.UpdateSubTaskRequest) (*pb.CommonWorkerResponse, error) {
+func (s *Server) UpdateSubTask(ctx context.Context, req *pb.UpdateSubTaskRequest) (*pb.OperateSubTaskResponse, error) {
 	log.Infof("[server] receive UpdateSubTask request %+v", req)
-
 	cfg := config.NewSubTaskConfig()
 	err := cfg.Decode(req.Task)
 	if err != nil {
-		log.Errorf("[server] decode config from request %+v error %v", req.Task, errors.ErrorStack(err))
-		return &pb.CommonWorkerResponse{
-			Result: false,
-			Msg:    errors.ErrorStack(err),
-		}, nil
+		err = errors.Annotatef(err, "decode config from request %+v", req.Task)
+		log.Errorf("[server] %s", errors.ErrorStack(err))
+		return nil, err
 	}
 
-	if err = s.worker.meta.Set(cfg); err != nil {
-		errMsg := fmt.Sprintf("[server] update task %s into meta: %v", cfg, errors.ErrorStack(err))
-		log.Errorf(errMsg)
-		return &pb.CommonWorkerResponse{
-			Result: false,
-			Msg:    errMsg,
-		}, nil
-	}
+	opLogID, err := s.worker.UpdateSubTask(cfg)
+	if err != nil {
+		err = errors.Annotatef(err, "update sub task %s", cfg.Name)
+		log.Errorf("[server] %s", errors.ErrorStack(err))
+		return nil, err
+	}
+
+	return &pb.OperateSubTaskResponse{
+		Meta: &pb.CommonWorkerResponse{
+			Result: true,
+			Msg:    "",
+		},
+		Op:    pb.TaskOp_Update,
+		LogID: opLogID,
+	}, nil
+}
 
-	err = s.worker.UpdateSubTask(cfg)
+// QueryTaskOperation implements WorkerServer.QueryTaskOperation
+func (s *Server) QueryTaskOperation(ctx context.Context, req *pb.QueryTaskOperationRequest) (*pb.QueryTaskOperationResponse, error) {
+	taskName := req.Name
+	opLogID := req.LogID
+
+	opLog, err := s.worker.meta.GetTaskLog(opLogID)
 	if err != nil {
-		log.Errorf("[server] update sub task %s error %v", cfg.Name, errors.ErrorStack(err))
-		return &pb.CommonWorkerResponse{
-			Result: false,
-			Msg:    errors.ErrorStack(err),
-		}, nil
+		log.Errorf("fail to get operation log %d of task %s", opLogID, taskName)
```

Comment:
use `errors.Annotatef`?

Reply:
lol, we should have a standard: how to add log in every caller